storage/shard.go

```diff
@@ -826,22 +779,23 @@ func (s *dbShard) writeAndIndex(
 	)
 	if writable {
 		// Perform write
-		err = entry.series.Write(ctx, timestamp, value, unit, annotation)
+		err = entry.Series.Write(ctx, timestamp, value, unit, annotation)
 		// Load series metadata before decrementing the writer count
 		// to ensure this metadata is snapshotted at a consistent state
 		// NB(r): We explicitly do not place the series ID back into a
 		// pool as high frequency users of series IDs such
 		// as the commit log need to use the reference without the
 		// overhead of ownership tracking. This makes taking a ref here safe.
-		commitLogSeriesID = entry.series.ID()
-		commitLogSeriesTags = entry.series.Tags()
-		commitLogSeriesUniqueIndex = entry.index
-		needsIndex := shouldReverseIndex && entry.needsIndexUpdate(timestamp)
-		if err == nil && needsIndex {
-			entry.onIndexPrepare()
-			err = s.reverseIndex.Write(entry.series.ID(), entry.series.Tags(), timestamp, entry)
-		}
-		entry.decrementReaderWriterCount()
+		commitLogSeriesID = entry.Series.ID()
+		commitLogSeriesTags = entry.Series.Tags()
+		commitLogSeriesUniqueIndex = entry.Index
+		if err == nil && shouldReverseIndex {
+			if entry.NeedsIndexUpdate(s.reverseIndex.BlockStartForWriteTime(timestamp)) {
```

Comment:
maybe add a comment that `NeedsIndexUpdate` has CAS semantics here so we don't change in an incompatible way later.